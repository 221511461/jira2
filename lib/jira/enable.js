const Installation = require('../models/installation')
const getAxiosInstance = require('../jira/client/axios')

module.exports = async (req, res) => {
  const installation = await Installation.getPendingHost(req.body.baseUrl)
  if (installation) {
    res.on('finish', async () => {
      const instance = getAxiosInstance(null, {}, installation.jiraHost, installation.sharedSecret)
      try {
        // Validate the clientKey and the sharedSecret
        // We have to do this after responding with 204 because
        // otherwise we get an error "The add-on is not enabled"
        // no matter what we send
        const result = await instance.get(`/rest/devinfo/0.10/existsByProperties?fakeProperty=1`)
        if (result.status === 200) {
          req.log('App enabled on Jira')
          await installation.enable()
        } else {
          req.log('App not installed. Verification failed.')
        }
      } catch (err) {
        if (err.response.status === 401) {
          req.log('Jira does not recognize this installation. Deleting it.')
          await Installation.uninstall({ clientKey: req.body.clientKey })
        }
      }
    })
    return res.sendStatus(204)
  } else {
    req.log(`No pending installation found for ${req.body.baseUrl}`)
    return res.sendStatus(422)
  }
}
