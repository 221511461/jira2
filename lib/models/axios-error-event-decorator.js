const url = require('url');

/*
 * Adds request/response metadata to a Sentry event for an Axios error
 * To use, pass AxiosErrorEventDecorator.decorate to scope.addEventProcessor
 *
 * See https://docs.sentry.io/platforms/node/#eventprocessors
 */
class AxiosErrorEventDecorator {
  static decorate(event, hint) {
    const decorator = new AxiosErrorEventDecorator(event, hint);
    return decorator.decorate();
  }

  constructor(event, hint) {
    this.event = event;
    this.hint = hint;
  }

  get error() {
    return this.hint.originalException;
  }

  get response() {
    return this.error.response;
  }

  get request() {
    return this.response.request;
  }

  validError() {
    return this.error && this.response && this.request;
  }

  decorate() {
    if (!this.validError()) {
      return this.event;
    }

    this.event.extra.response = this.responseMetadata();
    this.event.extra.request = this.requestMetadata();
    this.event.fingerprint = this.generateFingerprint();

    return this.event;
  }

  requestMetadata() {
    const metadata = {
      method: this.request.method,
      path: this.request.path,
      host: this.request.getHeader('host'),
      headers: this.request.getHeaders(),
    };

    const requestBody = this.response.config.data;
    if (requestBody) {
      metadata.body = this.parseRequestBody(requestBody);
    }

    return metadata;
  }

  responseMetadata() {
    const metadata = {
      status: this.response.status,
      headers: this.response.headers,
    };

    if (this.response.data) {
      metadata.body = this.response.data.slice(0, 255);
    }

    return metadata;
  }

  generateFingerprint() {
    const { pathname } = url.parse(this.request.path);

    return [
      '{{ default }}',
      this.response.status,
      `${this.request.method} ${pathname}`,
    ];
  }

  /*
   * Parse JSON body, when present and valid, otherwise return unparsed body.
   */
  parseRequestBody(body) {
    if (this.isJsonRequest()) {
      try {
        return JSON.parse(body);
      } catch (error) {
        if (error.name !== 'SyntaxError') {
          throw error;
        }
      }
    }

    return body;
  }

  isJsonRequest() {
    return this.request.getHeader('content-type').startsWith('application/json');
  }
}

module.exports = AxiosErrorEventDecorator;
