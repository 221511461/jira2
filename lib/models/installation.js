const crypto = require('crypto');
const Sequelize = require('sequelize');
const EncryptedField = require('sequelize-encrypted');
const Subscription = require('./subscription');

if (!process.env.STORAGE_SECRET) {
  throw new Error('STORAGE_SECRET is not defined.');
}

const encrypted = EncryptedField(Sequelize, process.env.STORAGE_SECRET);

function getHashedKey(clientKey) {
  const keyHash = crypto.createHmac('sha256', process.env.STORAGE_SECRET);
  keyHash.update(clientKey);

  return keyHash.digest('hex');
}

class Installation extends Sequelize.Model {
  static init(sequelize, DataTypes) {
    return super.init({
      jiraHost: DataTypes.STRING,
      secrets: encrypted.vault('secrets'),
      sharedSecret: encrypted.field('sharedSecret', {
        type: DataTypes.STRING,
        allowNull: false,
      }),
      clientKey: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      enabled: Sequelize.BOOLEAN,
    }, { sequelize });
  }

  static getForClientKey(clientKey) {
    return Installation.findOne({
      where: {
        clientKey: getHashedKey(clientKey),
      },
    });
  }

  static getForHost(host) {
    return Installation.findOne({
      where: {
        jiraHost: host,
        enabled: true,
      },
    });
  }

  static getPendingHost(jiraHost) {
    return Installation.findOne({
      where: {
        jiraHost,
        enabled: false,
      },
    });
  }

  async enable() {
    try {
      await this.update({
        enabled: true,
      });
    } catch (err) {
      logger.error(`Enable error: ${err}`);
    }
  }

  async disable() {
    try {
      await this.update({
        enabled: false,
      });
    } catch (err) {
      logger.error(`Disable error: ${err}`);
    }
  }

  /**
   * Create a new Installation object from a Jira Webhook
   *
   * @param {{host: string, clientKey: string, secret: string}} payload
   * @returns {Installation}
   */
  static async install(payload) {
    const [installation, created] = await Installation.findOrCreate({
      where: {
        clientKey: getHashedKey(payload.clientKey),
      },
      defaults: {
        jiraHost: payload.host,
        sharedSecret: payload.sharedSecret,
      },
    });

    if (!created) {
      try {
        await installation.update({
          sharedSecret: payload.sharedSecret,
          enabled: false,
          jiraHost: payload.host,
        }).then(async (record) => {
          const subscriptions = await Subscription.getAllForClientKey(record.clientKey);
          await Promise.all(subscriptions.map(subscription => subscription.update({ jiraHost: record.jiraHost })));

          return installation;
        });
      } catch (err) {
        logger.error(`Error updating subscription: ${err}`);
      }
    }

    try {
      await installation.update({
        enabled: false,
      });

      return installation;
    } catch (err) {
      logger.error(`Error updating installation: ${err}`);
    }
  }

  async uninstall() {
    this.destroy();
  }

  async subscriptions() {
    return Subscription.getAllForClientKey(this.clientKey);
  }
}

module.exports = Installation;
module.exports.getHashedKey = getHashedKey;
