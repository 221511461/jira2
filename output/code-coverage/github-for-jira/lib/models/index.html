
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for github-for-jira/lib/models</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> github-for-jira/lib/models</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">64.67% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>108/167</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">59.46% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>22/37</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">51.61% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>32/62</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">64.67% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>108/167</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="axios-error-event-decorator.js"><a href="axios-error-event-decorator.js.html">axios-error-event-decorator.js</a></td>
	<td data-value="94.12" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 94%"></div><div class="cover-empty" style="width: 6%"></div></div>
	</td>
	<td data-value="94.12" class="pct high">94.12%</td>
	<td data-value="34" class="abs high">32/34</td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="12" class="abs high">12/12</td>
	<td data-value="94.12" class="pct high">94.12%</td>
	<td data-value="34" class="abs high">32/34</td>
	</tr>

<tr>
	<td class="file high" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="83.33" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 83%"></div><div class="cover-empty" style="width: 17%"></div></div>
	</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="18" class="abs high">15/18</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="6" class="abs medium">3/6</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="3" class="abs low">1/3</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="18" class="abs high">15/18</td>
	</tr>

<tr>
	<td class="file low" data-value="installation.js"><a href="installation.js.html">installation.js</a></td>
	<td data-value="41.38" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 41%"></div><div class="cover-empty" style="width: 59%"></div></div>
	</td>
	<td data-value="41.38" class="pct low">41.38%</td>
	<td data-value="29" class="abs low">12/29</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="4" class="abs low">1/4</td>
	<td data-value="16.67" class="pct low">16.67%</td>
	<td data-value="12" class="abs low">2/12</td>
	<td data-value="41.38" class="pct low">41.38%</td>
	<td data-value="29" class="abs low">12/29</td>
	</tr>

<tr>
	<td class="file high" data-value="jira-client.js"><a href="jira-client.js.html">jira-client.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file high" data-value="octokit-error.js"><a href="octokit-error.js.html">octokit-error.js</a></td>
	<td data-value="95.83" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 95%"></div><div class="cover-empty" style="width: 5%"></div></div>
	</td>
	<td data-value="95.83" class="pct high">95.83%</td>
	<td data-value="24" class="abs high">23/24</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="95.83" class="pct high">95.83%</td>
	<td data-value="24" class="abs high">23/24</td>
	</tr>

<tr>
	<td class="file low" data-value="project.js"><a href="project.js.html">project.js</a></td>
	<td data-value="25" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 25%"></div><div class="cover-empty" style="width: 75%"></div></div>
	</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="12" class="abs low">3/12</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="16.67" class="pct low">16.67%</td>
	<td data-value="6" class="abs low">1/6</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="12" class="abs low">3/12</td>
	</tr>

<tr>
	<td class="file high" data-value="sentry-scope-proxy.js"><a href="sentry-scope-proxy.js.html">sentry-scope-proxy.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	</tr>

<tr>
	<td class="file low" data-value="subscription.js"><a href="subscription.js.html">subscription.js</a></td>
	<td data-value="10" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 10%"></div><div class="cover-empty" style="width: 90%"></div></div>
	</td>
	<td data-value="10" class="pct low">10%</td>
	<td data-value="30" class="abs low">3/30</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="7.14" class="pct low">7.14%</td>
	<td data-value="14" class="abs low">1/14</td>
	<td data-value="10" class="pct low">10%</td>
	<td data-value="30" class="abs low">3/30</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Thu Apr 01 2021 15:06:52 GMT+1100 (Australian Eastern Daylight Time)
            </div>
        </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    