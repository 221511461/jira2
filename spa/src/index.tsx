import React from "react";
import ReactDOM from "react-dom/client";
import {
	BrowserRouter,
	Route,
	Routes,
} from "react-router-dom";
import StartConnection from "./pages/StartConnection";
import "@atlaskit/css-reset";
import ConfigSteps from "./pages/ConfigSteps";

ReactDOM.createRoot(document.getElementById("root") as HTMLElement).render(
	<React.StrictMode>
		<BrowserRouter>
			<Routes>
				<Route path="/spa">
					<Route index element={<StartConnection />} />
					<Route path="steps" element={<ConfigSteps />} />
				</Route>
			</Routes>
		</BrowserRouter>
	</React.StrictMode>
);
