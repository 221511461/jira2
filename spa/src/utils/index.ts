import * as Sentry from "@sentry/react";
import { AxiosError } from "axios";
import { BackfillStatusResp, GHSubscriptions } from "../rest-interfaces";

export const getJiraJWT = (): Promise<string> => new Promise(resolve => {
	return AP.context.getToken((token: string) => {
		resolve(token);
	});
});

export function popup (url: string) {
	const openedPopup = window.open(url, "_blank");
	if (!openedPopup || openedPopup.closed || typeof openedPopup.closed === "undefined") {
		return null;
	}
	return openedPopup;
}

export function reportError(err: unknown, extra: {
	path: string,
	reason?: string
} & Record<string, unknown>) {
	try {

		const cause = (err as Record<string, unknown>).cause || {};
		delete (err as Record<string, unknown>).cause; //so that Sentry doesn't group all axios error together

		Sentry.captureException(err, {
			extra: {
				...extra,
				...(err instanceof AxiosError ? extractKeyErrorInfo(err) : {}),
				cause: {
					...(cause instanceof AxiosError ? extractKeyErrorInfo(cause) : cause),
				}
			}
		});
	} catch (_) {
		//do nothing
	}
}

function extractKeyErrorInfo(e: AxiosError) {
	return {
		errMessage: e.message,
		errCode: e.code,
		errMethod: e.config?.method,
		errStatusCode: e.response?.status,
		errBody: e.response?.data
	};
}

export function openChildWindow(url: string) {
	const child: Window | null = window.open(url);
	const interval = setInterval(function () {
		if (child?.closed) {
			clearInterval(interval);
			AP.navigator.reload();
		}
	}, 100);
	return child;
}

export const getInProgressSubIds = (response: GHSubscriptions): Array<number> => {
	const successfulCloudConnections =
		response.ghCloudSubscriptions.successfulCloudConnections;
	const inProgressCloudConnections = successfulCloudConnections.filter(
		(connection) =>
			connection.syncStatus === "IN PROGRESS" ||
			connection.syncStatus === "PENDING"
	);
	const inProgressCloudSubIds = inProgressCloudConnections.map(
		(connection) => connection.subscriptionId
	);
	let inProgressGHESubIds: Array<number> = [];
	const ghEnterpriseServers = response.ghEnterpriseServers;
	for (const ghEnterpriseServer of ghEnterpriseServers) {
		const applications = ghEnterpriseServer.applications;
		for (const application of applications) {
			const successfulGHEConnections = application.successfulConnections;
			const inProgressGHEConnections = successfulGHEConnections.filter(
				(connection) =>
					connection.syncStatus === "IN PROGRESS" ||
					connection.syncStatus === "PENDING"
			);
			inProgressGHESubIds = inProgressGHEConnections.map(
				(connection) => connection.subscriptionId
			);
		}
	}
	return [...inProgressCloudSubIds, ...inProgressGHESubIds];
};

export const getUpdatedSubscriptions = (response: BackfillStatusResp, subscriptions: GHSubscriptions): GHSubscriptions | undefined => {
	const currentSubs = subscriptions;
	let newSubs;
	const subscriptionIds = response.subscriptionIds || [];
	if (currentSubs) {
		for (const subscriptionId of subscriptionIds) {
			const subscriptions = response.subscriptions;
			const subscription = subscriptions[subscriptionId];
			if (!subscription.gitHubAppId) {
				let matchedIndex;
				const successfulCloudConnections =
					currentSubs?.ghCloudSubscriptions.successfulCloudConnections;
				if (successfulCloudConnections) {
					matchedIndex = successfulCloudConnections.findIndex(
						(connection) => connection.subscriptionId === subscriptionId
					);
					successfulCloudConnections[matchedIndex] = {
						...successfulCloudConnections[matchedIndex],
						numberOfSyncedRepos: subscription.syncedRepos,
						syncStatus: subscription.syncStatus,
					};
					if(subscription.backfillSince){
						successfulCloudConnections[matchedIndex]["backfillSince"] = subscription.backfillSince;
					}
					newSubs = {
						...currentSubs,
						ghCloudSubscriptions: {
							...currentSubs.ghCloudSubscriptions,
							successfulCloudConnections: successfulCloudConnections,
						},
					};
				}
			} else {
				const ghEnterpriseServers = currentSubs.ghEnterpriseServers;
				let ghEnterpriseServerIndex;
				let applicationIndex;
				for (const [
					ghEnterpriseServerI,
					ghEnterpriseServer,
				] of ghEnterpriseServers.entries()) {
					const applications = ghEnterpriseServer.applications;
					for (const [applicationI, app] of applications.entries()) {
						if (app.id === subscription.gitHubAppId) {
							ghEnterpriseServerIndex = ghEnterpriseServerI;
							applicationIndex = applicationI;
							break;
						}
					}
				}
				console.log("::::: ghEnterpriseServerIndex",ghEnterpriseServerIndex);
				console.log("::::: applicationIndex", applicationIndex);
				if (ghEnterpriseServerIndex && applicationIndex) {
					const newGHEnterpriseServers = ghEnterpriseServers;
					const newApps =
						ghEnterpriseServers[ghEnterpriseServerIndex].applications;
					newApps[applicationIndex] = {
						...newApps[applicationIndex],
						successfulConnections: [],
					};
					newGHEnterpriseServers[ghEnterpriseServerIndex] = {
						...ghEnterpriseServers[ghEnterpriseServerIndex],
						applications: newApps,
					};
					newSubs = {
						...currentSubs,
						ghEnterpriseServers: newGHEnterpriseServers,
					};
				}
			}
		}
		return newSubs;
	}
};
