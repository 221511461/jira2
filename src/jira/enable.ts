import {ActionFromInstallation, ActionSource, ActionType,} from '../proto/v0/action';
import {submitProto} from '../tracking';
import {Installation} from '../models';
import verifyInstallation from './verify-installation';
import {Request, Response} from 'express';

/**
 * Handle the enable webhook from Jira
 */
export default async (req: Request, res: Response): Promise<void> => {
  const jiraHost = req.body.baseUrl;

  const installation = await Installation.getPendingHost(jiraHost);
  if (!installation) {
    req.log.info(`No pending installation found for jiraHost=${jiraHost}`);
    res.sendStatus(422);
    return;
  }

  const action = await ActionFromInstallation(installation);
  action.type = ActionType.ENABLED;
  action.actionSource = ActionSource.WEBHOOK;
  res.on('finish', verifyInstallation(installation, req.log));
  res.sendStatus(204);
  await submitProto(action);
};
