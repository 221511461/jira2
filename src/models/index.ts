import { logger } from 'probot/lib/logger';
import InstallationModel from './installation';
import SubscriptionModel from './subscription';
import ProjectModel from './project';
import Sequelize, { DataTypes } from 'sequelize';
import EncryptedField from 'sequelize-encrypted';

// TODO: need to move this into a function
if (!process.env.STORAGE_SECRET) {
  throw new Error('STORAGE_SECRET is not defined.');
}

const encrypted = EncryptedField(Sequelize, process.env.STORAGE_SECRET);

// TODO: config misses timezone config to force to UTC, defaults to local timezone of PST
// eslint-disable-next-line @typescript-eslint/no-var-requires
const config = require('../../db/config.json')[
  process.env.NODE_ENV || 'development'
];

config.benchmark = true;
config.logging = config.disable_sql_logging
  ? undefined
  : (query, ms) => logger.debug({ ms }, query);

export const sequelize = config.use_env_variable
  ? new Sequelize.Sequelize(process.env[config.use_env_variable], config)
  : new Sequelize.Sequelize(
      config.database,
      config.username,
      config.password,
      config,
    );

InstallationModel &&
  InstallationModel.init(
    {
      jiraHost: DataTypes.STRING,
      secrets: encrypted.vault('secrets'),
      sharedSecret: encrypted.field('sharedSecret', {
        type: DataTypes.STRING,
        allowNull: false,
      }),
      clientKey: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      enabled: Sequelize.BOOLEAN,
    },
    { sequelize },
  );

SubscriptionModel &&
  SubscriptionModel.init(
    {
      gitHubInstallationId: DataTypes.INTEGER,
      jiraHost: DataTypes.STRING,
      selectedRepositories: DataTypes.ARRAY(DataTypes.INTEGER),
      repoSyncState: DataTypes.JSONB,
      syncStatus: DataTypes.ENUM('PENDING', 'COMPLETE', 'ACTIVE', 'FAILED'),
      syncWarning: DataTypes.STRING,
      jiraClientKey: DataTypes.STRING,
    },
    { sequelize },
  );

ProjectModel.init(
  {
    projectKey: DataTypes.STRING,
    occurrences: DataTypes.INTEGER,
    jiraHost: DataTypes.STRING,
  },
  { sequelize },
);

export const Installation = InstallationModel;
export const Subscription = SubscriptionModel;
export const Project = ProjectModel;
