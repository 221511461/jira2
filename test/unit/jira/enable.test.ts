/* eslint-disable @typescript-eslint/no-explicit-any */
import enable from "../../../src/jira/enable";
import { Installation } from "../../../src/models";
import { mocked } from "ts-jest/utils";


jest.mock("../../../src/models");

describe("Webhook: /events/enabled", () => {
	let installation;

	beforeEach(async () => {
		installation = {
			id: 19,
			jiraHost: "https://test-host.jira.com",
			clientKey: "abc123",
			enabled: true,
			secrets: "def234",
			sharedSecret: "ghi345",
			subscriptions: jest.fn().mockResolvedValue([])
		};

		// Allows us to modify installation before it's finally called
		mocked(Installation.getPendingHost).mockImplementation(() => installation);

	});

	it("Pending Installation", async () => {
		const req = {
			log: { info: jest.fn() },
			body: { baseUrl: installation.jiraHost }
		};
		const res = { sendStatus: jest.fn(), on: jest.fn() };

		await enable(req as any, res as any);
		expect(res.on).toHaveBeenCalledWith("finish", expect.any(Function));
		expect(res.sendStatus).toHaveBeenCalledWith(204);
	});

	it("No Pending Installation", async () => {
		installation = null;
		const req = {
			log: { info: jest.fn() },
			body: { baseUrl: "https://no-exist.jira.com" }
		};
		const res = { sendStatus: jest.fn(), on: jest.fn() };

		await enable(req as any, res as any);
		expect(res.sendStatus).toHaveBeenCalledWith(422);
	});
});
