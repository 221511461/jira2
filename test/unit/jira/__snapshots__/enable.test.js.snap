// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhook: /events/enabled Pending Installation 1`] = `"jira.v0.Action"`;

exports[`Webhook: /events/enabled Pending Installation 2`] = `
Object {
  "actionSource": "webhook",
  "association": "installation",
  "installationId": 19,
  "jiraHostname": "https://test-host.jira.com",
  "type": "enabled",
}
`;
